File: Day22/Part1/strange_printer.py
# LC 664 — Strange Printer
# Interval DP: dp[i][j] = min turns to print s[i..j]

from functools import lru_cache

def strangePrinter(s: str) -> int:
    if not s:
        return 0
    # compress consecutive duplicates
    t = []
    for ch in s:
        if not t or t[-1] != ch:
            t.append(ch)
    s = ''.join(t)
    n = len(s)

    @lru_cache(None)
    def dp(i: int, j: int) -> int:
        if i > j:
            return 0
        if i == j:
            return 1
        res = dp(i+1, j) + 1
        for k in range(i+1, j+1):
            if s[i] == s[k]:
                res = min(res, dp(i+1, k-1) + dp(k, j))
        return res

    return dp(0, n-1)

if __name__ == "__main__":
    print(strangePrinter("aba"))  # 2

File: learnLeetCode/Day22/Part1/strange_printer.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 664 — Strange Printer
// Interval DP with compression of consecutive duplicates

int strangePrinter(string s){
    string t; t.reserve(s.size());
    for(char c: s){ if(t.empty() || t.back()!=c) t.push_back(c); }
    int n = (int)t.size();
    if(n==0) return 0;
    vector<vector<int>> dp(n, vector<int>(n, 0));
    for(int i=0;i<n;++i) dp[i][i]=1;
    for(int len=2; len<=n; ++len){
        for(int i=0; i+len-1<n; ++i){
            int j = i+len-1;
            int res = dp[i+1][j] + 1;
            for(int k=i+1; k<=j; ++k){
                if(t[i]==t[k]){
                    int left = (k-1>=i+1) ? dp[i+1][k-1] : 0;
                    res = min(res, left + dp[k][j]);
                }
            }
            dp[i][j] = res;
        }
    }
    return dp[0][n-1];
}

int main(){
    cout << strangePrinter("aba") << "\n"; // 2
    return 0;
}
